File: packages/vscode/src/mcp-toml.ts
import { join } from "node:path"

export const MCP_SERVER_TABLE = "mcp_servers.groks-beard"

export const TUI_BRIDGE_REFRESH_MESSAGE =
  "TUI bridge enabled. In a running TUI press r in /mcps, or start a new TUI session. The server does not attach live."

export const projectGrokConfigPath = (workspaceRoot: string): string =>
  join(workspaceRoot, ".grok", "config.toml")

const tomlString = (value: string): string => JSON.stringify(value)

export const renderMcpTable = (
  nodeCommand: string,
  proxyPath: string,
  workspace: string,
): string =>
  `[${MCP_SERVER_TABLE}]\ncommand = ${tomlString(nodeCommand)}\nargs = [${
    tomlString(proxyPath)
  }, "--workspace", ${tomlString(workspace)}]\n`

const escapeRegExp = (value: string): string => value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")

const tableRange = (text: string, header: string): { start: number; end: number } | undefined => {
  const re = new RegExp(`^\\[${escapeRegExp(header)}\\][ \\t]*\\r?\\n?`, "m")
  const match = re.exec(text)
  if (match === null) return undefined
  const start = match.index
  const after = start + match[0].length
  const rest = text.slice(after)
  const next = /^[ \t]*\[/m.exec(rest)
  const end = next === null ? text.length : after + next.index
  return { start, end }
}

export const mergeMcpTable = (existing: string, table: string): string => {
  const range = tableRange(existing, MCP_SERVER_TABLE)
  if (range === undefined) {
    const trimmed = existing.replace(/\s*$/, "")
    return trimmed === "" ? table : `${trimmed}\n\n${table}`
  }
  const before = existing.slice(0, range.start).replace(/\s*$/, "")
  const after = existing.slice(range.end).replace(/^\s*/, "")
  const parts = [before, table.replace(/\s*$/, ""), after].filter((part) => part.length > 0)
  return `${parts.join("\n\n")}\n`
}

export const removeMcpTable = (existing: string): string => {
  const range = tableRange(existing, MCP_SERVER_TABLE)
  if (range === undefined) return existing
  const before = existing.slice(0, range.start).replace(/\s*$/, "")
  const after = existing.slice(range.end).replace(/^\s*/, "")
  if (before === "") return after === "" ? "" : `${after}${after.endsWith("\n") ? "" : "\n"}`
  if (after === "") return `${before}\n`
  return `${before}\n\n${after}${after.endsWith("\n") ? "" : "\n"}`
}

File: beard/core/src/main/scala/groksbeard/core/McpToml.scala
package groksbeard.core

object McpToml:
  val ServerTable       = "mcp_servers.groks-beard"
  val EditorDownMessage =
    "Open this workspace in VS Code or Cursor with Grok's Beard and run \"Grok's Beard: Enable TUI Bridge\"."
  val RefreshMessage =
    "TUI bridge enabled. In a running TUI press r in /mcps, or start a new TUI session. The server does not attach live."

  def projectConfigPath(workspaceRoot: String): String =
    val root = workspaceRoot.stripSuffix("/").stripSuffix("\\")
    s"$root/.grok/config.toml"

  def renderTable(nodeCommand: String, proxyPath: String, workspace: String): String =
    val cmd  = jsonString(nodeCommand)
    val prox = jsonString(proxyPath)
    val ws   = jsonString(workspace)
    s"[$ServerTable]\ncommand = $cmd\nargs = [$prox, \"--workspace\", $ws]\n"

  def mergeTable(existing: String, table: String): String =
    tableRange(existing, ServerTable) match
      case None =>
        val trimmed = existing.replaceAll("\\s+$", "")
        if trimmed.isEmpty then table else s"$trimmed\n\n$table"
      case Some((start, end)) =>
        val before = existing.substring(0, start).replaceAll("\\s+$", "")
        val after  = existing.substring(end).replaceAll("^\\s+", "")
        val parts  = List(before, table.replaceAll("\\s+$", ""), after).filter(_.nonEmpty)
        parts.mkString("\n\n") + "\n"

  def removeTable(existing: String): String =
    tableRange(existing, ServerTable) match
      case None               => existing
      case Some((start, end)) =>
        val before = existing.substring(0, start).replaceAll("\\s+$", "")
        val after  = existing.substring(end).replaceAll("^\\s+", "")
        if before.isEmpty then if after.isEmpty then "" else if after.endsWith("\n") then after else s"$after\n"
        else if after.isEmpty then s"$before\n"
        else s"$before\n\n${if after.endsWith("\n") then after else s"$after\n"}"

  private def jsonString(value: String): String =
    val escaped = value.flatMap {
      case '"'  => "\\\""
      case '\\' => "\\\\"
      case c    => c.toString
    }
    s"\"$escaped\""

  private def tableRange(text: String, header: String): Option[(Int, Int)] =
    val escaped = header.replaceAll("([.*+?^${}()|\\[\\]\\\\])", "\\\\$1")
    val re      = s"(?m)^\\[$escaped\\][ \\t]*\\r?\\n?".r
    re.findFirstMatchIn(text).map { m =>
      val start = m.start
      val after = m.end
      val rest  = text.substring(after)
      val next  = "(?m)^[ \\t]*\\[".r.findFirstMatchIn(rest)
      val end   = next.map(n => after + n.start).getOrElse(text.length)
      (start, end)
    }
  end tableRange
end McpToml
